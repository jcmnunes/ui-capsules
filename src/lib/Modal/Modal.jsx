import React from 'react';
import PropTypes from 'prop-types';
import ReactModal from 'react-modal';
import styled from 'styled-components';
import Button from '../Button/Button';
import theme from '../theme';

const StyledModal = styled(ReactModal)`
  position: absolute;
  top: 50%;
  left: 16px;
  right: 16px;
  bottom: auto;
  transform: translate(0, -50%);
  background: #fff;
  overflow: auto;
  -webkit-overflow-scrolling: touch;
  outline: none;
  width: auto;
  min-width: 300px;
  box-shadow: rgba(0, 0, 0, 0.15) 0 10px 20px, rgba(0, 0, 0, 0.1) 0 3px 6px;
  border-width: 6px;
  border-style: solid;
  border-color: ${theme.neutral200};
  border-image: initial;
  z-index: 999999;

  @media (min-width: ${theme.bp_mobile}) {
    left: 40px;
    right: 40px;
  }

  @media (min-width: ${theme.bp_small}) {
    top: 50%;
    left: 50%;
    right: auto;
    bottom: auto;
    margin-right: -50%;
    width: ${props => props.width || '640px'};
    transform: translate(-50%, -50%);
  }
`;

const ModalHeader = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-end;
  height: 40px;
  padding: 16px;
  color: ${theme.neutral200};
`;

export const ModalBody = styled.div`
  padding: 12px 32px 32px;
`;

export const ModalTitle = styled.h2`
  font-size: 24px;
  margin-bottom: 16px;
  color: inherit;
`;

export const ModalFooter = styled.div`
  display: flex;
  justify-content: flex-end;
  background: ${theme.neutral050};
  padding: 18px 24px;

  & > * {
    margin-left: 16px;
  }
`;

const Modal = ({ isOpen, onRequestClose, contentLabel, width, children, ...other }) => (
  <StyledModal
    isOpen={isOpen}
    onRequestClose={onRequestClose}
    contentLabel={contentLabel}
    width={width}
    closeTimeoutMS={200}
    {...other}
  >
    <ModalHeader>
      {!!onRequestClose && (
        <Button appearance="none" size="medium" iconBefore="CLOSE" onClick={onRequestClose} />
      )}
    </ModalHeader>
    {children}
  </StyledModal>
);
Modal.displayName = 'Modal';

Modal.defaultProps = {
  width: '640px',
  onRequestClose: null,
};

Modal.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  contentLabel: PropTypes.string.isRequired,
  onRequestClose: PropTypes.func,
  width: PropTypes.string,
  children: PropTypes.node.isRequired,
};

export default Modal;
export { ReactModal };
