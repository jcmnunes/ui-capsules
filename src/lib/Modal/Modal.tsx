import React from 'react';
import ReactModal from 'react-modal';
import styled from 'styled-components';
import { Button } from '../Button/Button';
import { theme } from '../theme';

const styles = {
  overlay: {
    backgroundColor: 'rgba(60, 63, 75, 0.65)',
  },
};

interface StyledModalProps {
  width?: string;
}
const StyledModal = styled(ReactModal)<StyledModalProps>`
  position: absolute;
  top: 50%;
  left: 16px;
  right: 16px;
  bottom: auto;
  transform: translate(0, -50%);
  background: #fff;
  overflow: auto;
  -webkit-overflow-scrolling: touch;
  outline: none;
  width: auto;
  min-width: 300px;
  box-shadow: rgba(0, 0, 0, 0.15) 0 10px 20px, rgba(0, 0, 0, 0.1) 0 3px 6px;
  border: none;
  border-radius: 16px;
  border-image: initial;
  z-index: ${theme.modalZIndex};

  @media (min-width: ${theme.breakpoints.mobile}) {
    left: 40px;
    right: 40px;
  }

  @media (min-width: ${theme.breakpoints.sm}) {
    top: 50%;
    left: 50%;
    right: auto;
    bottom: auto;
    margin-right: -50%;
    width: ${props => props.width || '640px'};
    transform: translate(-50%, -50%);
  }
`;

const ModalHeader = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-end;
  height: 40px;
  padding: 16px 16px 0;
  color: ${theme.neutral200};
`;

export const ModalBody = styled.div`
  padding: 32px 32px 48px;
`;

export const ModalTitle = styled.h2`
  font-size: 24px;
  margin-bottom: 16px;
  color: inherit;
`;

export const ModalFooter = styled.div`
  display: flex;
  justify-content: flex-end;
  background: ${theme.neutral050};
  padding: 18px 24px;

  & > * {
    margin-left: 16px;
  }
`;

interface Props {
  isOpen: boolean;
  contentLabel: string;
  onRequestClose?(): void;
  width?: string;
  showCloseButton?: boolean;
}

export const Modal: React.FC<Props> = ({
  isOpen,
  onRequestClose,
  contentLabel,
  width,
  showCloseButton,
  children,
  ...other
}) => (
  <StyledModal
    isOpen={isOpen}
    onRequestClose={onRequestClose}
    contentLabel={contentLabel}
    width={width}
    closeTimeoutMS={200}
    style={styles}
    {...other}
  >
    <ModalHeader>
      {!!onRequestClose && showCloseButton && (
        <Button appearance="none" size="medium" iconBefore="X" onClick={onRequestClose} />
      )}
    </ModalHeader>
    {children}
  </StyledModal>
);
Modal.displayName = 'Modal';

Modal.defaultProps = {
  width: '640px',
  onRequestClose: undefined,
  showCloseButton: true,
};

export { ReactModal };
