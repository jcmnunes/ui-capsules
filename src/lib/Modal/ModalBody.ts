import styled from '@emotion/styled/macro';

export const ModalBody = styled.div`
  padding: 0 32px 24px;
`;
