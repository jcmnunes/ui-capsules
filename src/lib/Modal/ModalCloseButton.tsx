import React, { FC, MouseEvent } from 'react';
import { IconButton } from '../IconButton/IconButton';
import { styled } from 'styled-components';

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
const StyledIconButton = styled(IconButton)({
  position: 'absolute !important',
  top: 24,
  right: 24,

  '&:hover': {
    background: 'transparent',
  },
});

export interface Props {
  size?: 'small' | 'medium';
  onClick(e: MouseEvent<HTMLButtonElement>): void;
}

export const ModalCloseButton: FC<Props> = ({ size, onClick }) => {
  return (
    <StyledIconButton
      variant="ghostGray"
      size={size}
      icon="x"
      onClick={onClick}
      aria-label="Close"
    />
  );
};
