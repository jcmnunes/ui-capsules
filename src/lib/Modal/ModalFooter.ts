import { styled } from '../stitches.config';

export const ModalFooter = styled('div', {
  display: 'flex',
  justifyContent: 'flex-end',
  background: '$bg',
  padding: '16px 24px',

  '& > *': {
    marginLeft: 8,
  },
});
