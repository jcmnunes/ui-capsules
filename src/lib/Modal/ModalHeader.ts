import styled from '@emotion/styled/macro';

export const ModalHeader = styled.div(({ theme }) => ({
  display: 'flex',
  flexDirection: 'row',
  padding: '24px 64px 16px 32px',
  color: theme.colors.neutral['700'],
  fontSize: '22px',
  fontWeight: 500,
  lineHeight: '32px',
}));
