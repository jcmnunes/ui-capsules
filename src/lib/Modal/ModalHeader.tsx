import React from 'react';
import { Icon } from '../Icon/Icon';
import { darkTheme } from '../darkTheme';
import { IconType, Status } from '../types';
import { styled } from 'styled-components';

export const StyledModalHeader = styled.div(({ theme }) => ({
  display: 'flex',
  flexDirection: 'row',
  alignItems: 'center',
  padding: '24px 64px 4px 24px',
  color: theme.colors.neutral700,
  fontSize: theme.fontSizes['2xl'],
  fontWeight: 500,
  lineHeight: '32px',
}));

interface IconContainerProps {
  variant?: Status;
}

export const IconContainer = styled.div<IconContainerProps>(
  {
    width: 32,
    height: 32,
    borderRadius: 16,
    padding: 4,
    marginRight: 8,
  },

  ({ theme, variant }) => {
    switch (variant) {
      case 'success':
        return {
          background: theme.colors.success100,

          '.darkTheme &': {
            background: theme.colors.neutral100,
          },
        };
      case 'warning':
        return {
          background: theme.colors.warning100,

          '.darkTheme &': {
            background: theme.colors.neutral100,
          },
        };
      case 'error':
        return {
          background: theme.colors.error100,

          '.darkTheme &': {
            background: theme.colors.neutral100,
          },
        };
    }
  },
);

interface StyledIconProps {
  variant?: Status;
}

const StyledIcon = styled(Icon)<StyledIconProps>(({ theme, variant }) => {
  switch (variant) {
    case 'success':
      return {
        color: theme.colors.success700,

        '.darkTheme &': {
          color: theme.colors.success400,
        },
      };
    case 'warning':
      return {
        color: theme.colors.warning700,

        '.darkTheme &': {
          color: theme.colors.warning300,
        },
      };
    case 'error':
      return {
        color: theme.colors.error700,

        '.darkTheme &': {
          color: theme.colors.error400,
        },
      };
  }
});

interface ModalHeaderProps {
  variant?: Status;
}

const iconMap: Record<Status, IconType> = {
  success: 'check_c',
  warning: 'exclamation_t',
  error: 'exclamation_c',
};

export const ModalHeader: React.FC<ModalHeaderProps> = ({ variant, children }) => {
  const icon = variant ? iconMap[variant] : null;

  return (
    <StyledModalHeader>
      {icon && (
        <IconContainer variant={variant}>
          <StyledIcon icon={icon} variant={variant} size={24} />
        </IconContainer>
      )}

      {children}
    </StyledModalHeader>
  );
};
