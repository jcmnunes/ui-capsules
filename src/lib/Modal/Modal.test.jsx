import React from 'react';
import { render } from '@testing-library/react';
import RModal from 'react-modal';
import Modal, { ModalBody, ModalFooter, ModalTitle } from './Modal';

const bindApp = () => {
  const app = document.createElement('div');
  app.setAttribute('id', 'root');
  document.body.append(app);
  RModal.setAppElement('#root');
};

const CONTENT_LABEL = 'Example Modal';
const LABEL_TEXT = 'Label text';

const getModalElement = baseElement => baseElement.querySelector(`[aria-label="${CONTENT_LABEL}"]`);

const ModalContent = () => (
  <>
    <ModalBody>
      <ModalTitle>Modal title</ModalTitle>
      <label htmlFor="input-id">
        {LABEL_TEXT}
        <input id="input-id" type="text" />
      </label>
    </ModalBody>
    <ModalFooter>Footer</ModalFooter>
  </>
);

beforeEach(() => {
  bindApp();
});

afterEach(() => {
  document.body.removeChild(document.querySelector('#root'));
});

it('should have a displayName', () => {
  expect(Modal.displayName).toBe('Modal');
});

it('should render with no errors', () => {
  const { baseElement, getByLabelText } = render(
    <Modal isOpen contentLabel={CONTENT_LABEL} onRequestClose={() => {}}>
      <ModalContent />
    </Modal>,
  );
  expect(getByLabelText(LABEL_TEXT)).toBeInTheDocument();
  expect(getModalElement(baseElement)).toMatchSnapshot();
});

it('should not render the modal when isOpen is set to false', () => {
  const { baseElement } = render(
    <Modal isOpen={false} contentLabel={CONTENT_LABEL} onRequestClose={() => {}}>
      <ModalContent />
    </Modal>,
  );
  expect(getModalElement(baseElement)).toBeNull();
});
