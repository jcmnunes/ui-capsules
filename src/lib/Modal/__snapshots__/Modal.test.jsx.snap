// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render with no errors 1`] = `
.c3 {
  visibility: visible;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  position: relative;
  height: 32px;
  color: inherit;
  border-radius: 4px;
  border: none;
  border: none;
  font-size: 16px;
  font-family: 'Source Sans Pro',Helvetica,Arial,sans-serif;
  padding: 4px 6px;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  background: linear-gradient(to bottom, none, none );
}

.c2::-webkit-input-placeholder {
  color: #9AA5B1;
}

.c2::-moz-placeholder {
  color: #9AA5B1;
}

.c2:-ms-input-placeholder {
  color: #9AA5B1;
}

.c2::placeholder {
  color: #9AA5B1;
}

.c2:hover {
  background: linear-gradient(to bottom, none, none );
}

.c2:active {
  background: none;
}

.c2:disabled {
  background: none;
  cursor: not-allowed;
}

.c2:focus {
  outline: none;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0 {
  position: absolute;
  top: 50%;
  left: 16px;
  right: 16px;
  bottom: auto;
  -webkit-transform: translate(0,-50%);
  -ms-transform: translate(0,-50%);
  transform: translate(0,-50%);
  background: #fff;
  overflow: auto;
  -webkit-overflow-scrolling: touch;
  outline: none;
  width: auto;
  min-width: 300px;
  box-shadow: rgba(0,0,0,0.15) 0 10px 20px,rgba(0,0,0,0.1) 0 3px 6px;
  border-width: 6px;
  border-style: solid;
  border-color: #CBD2D8;
  border-image: initial;
  z-index: 999999;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  height: 40px;
  padding: 16px;
  color: #CBD2D8;
}

.c4 {
  padding: 12px 32px 32px;
}

.c5 {
  font-size: 24px;
  margin-bottom: 16px;
  color: inherit;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  background: #F5F7FA;
  padding: 18px 24px;
}

.c6 > * {
  margin-left: 16px;
}

@media (min-width:450px) {
  .c0 {
    left: 40px;
    right: 40px;
  }
}

@media (min-width:720px) {
  .c0 {
    top: 50%;
    left: 50%;
    right: auto;
    bottom: auto;
    margin-right: -50%;
    width: 640px;
    -webkit-transform: translate(-50%,-50%);
    -ms-transform: translate(-50%,-50%);
    transform: translate(-50%,-50%);
  }
}

<div
  aria-label="Example Modal"
  class="ReactModal__Content ReactModal__Content--after-open c0"
  role="dialog"
  tabindex="-1"
>
  <div
    class="c1"
  >
    <button
      class="c2"
      type="button"
    >
      <div
        class="c3"
      >
        <svg
          height="20px"
          viewBox="0 0 20 20"
          width="20px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11.88 10l5.73-5.71a1.34 1.34 0 1 0-1.9-1.9L10 8.12 4.29 2.39a1.34 1.34 0 0 0-1.9 1.9L8.12 10l-5.73 5.71a1.34 1.34 0 0 0 0 1.89 1.34 1.34 0 0 0 1.89 0L10 11.88l5.71 5.73a1.34 1.34 0 0 0 1.89 0 1.34 1.34 0 0 0 0-1.89z"
            fill="currentColor"
          />
        </svg>
      </div>
    </button>
  </div>
  <div
    class="c4"
  >
    <h2
      class="c5"
    >
      Modal title
    </h2>
    <label
      for="input-id"
    >
      Label text
      <input
        id="input-id"
        type="text"
      />
    </label>
  </div>
  <div
    class="c6"
  >
    Footer
  </div>
</div>
`;
